### PR TITLE
[#57] Initialize ligo contract

### PR DESCRIPTION
**Pull request by Martoon-00**
## Description

Stub for our LIGO version of the contract.

- [x] Decide on which syntax to use
- [x] Fill the most basic parts

## Related issue(s)

Initializes #57
Mostly solves #97 as well (except from the point about `:metadata`) 

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the following policy](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: ligo/Makefile
Comment (gromakovsky) [inline ligo/Makefile:18]: The way we do in `stablecoin` (where the target is an actual generated file, not a phony target) looks like a more proper usage of Makefiles to me, but I don't insist on doing that way.

File: ligo/Makefile
Comment (gromakovsky) [inline ligo/Makefile:13]: `trivialDAO` looks like a leftover.

File: ligo/Makefile
Comment (gromakovsky) [inline ligo/Makefile:8]: What syntax are we using? I recall this message in our chat:
> Regrading ligo syntax, let's pick Pascal. The tutorial is written on this dialect so using it should be simpler

File: ligo/Makefile
Comment (gromakovsky) [inline ligo/Makefile:26]: Why are we doing it?

File: .buildkite/pipeline.yml
Comment (gromakovsky) [inline .buildkite/pipeline.yml:62]: I think we should also put the resulting contract into artifacts.

File: ligo/test/Ligo/BaseDAO/Contract.hs
Comment (gromakovsky) [inline ligo/test/Ligo/BaseDAO/Contract.hs:32]: Can we use TH like we do in https://github.com/tqtezos/stablecoin? It sounds safer and cleaner. Maybe we should even add a helper to `morley` to import a contract in compile-time.

**Comment by Martoon-00** [comment]
I added a few small fixes to LIGO types + there are tests (which do not pass and are not checked though yet)

File: ligo/Makefile
Comment (Martoon-00) [reply ligo/Makefile:18]: Do you know how do they use their way? Just `make stablecoin.tz` (as in CI), or their approach allows something smarter?

File: ligo/Makefile
Comment (Martoon-00) [reply ligo/Makefile:8]: Eventually Kirill Kuvshinov opened my eyes and now I see that PascalLigo is not that nice, better to use camelligo.

File: ligo/Makefile
Comment (Martoon-00) [reply ligo/Makefile:26]: LIGO prints a contract as 

```
{ parameter ...;
  storage ...;
  code ...; }
```

This makes the contract more difficult to read (for instance I needed to do `diff baseDAOLorentz.tz baseDAOLigo.tz`, and it got broken because of the braces). Since I needed to implement this, probably it makes sense to leave there (though I don't like these extra lines being printed on each `make build`, not sure what to do with this).

File: .buildkite/pipeline.yml
Comment (Martoon-00) [reply .buildkite/pipeline.yml:62]: I tried hard but failed, will offload this to Ops team (hopefully after this PR is merged, because I don't want to block anyone).

File: ligo/test/Ligo/BaseDAO/Contract.hs
Comment (Martoon-00) [reply ligo/test/Ligo/BaseDAO/Contract.hs:32]: Apparently in this case this makes perfect sense.

File: ligo/Makefile
Comment (gromakovsky) [reply ligo/Makefile:8]: IIRC it was your lowest priority, but ok, I don't have a preference anyway. It means we won't be able to copy-paste FA2 implementation from Stablecoin, but ok, we will port it to a different LIGO syntax, shouldn't be hard.

File: ligo/Makefile
Comment (gromakovsky) [reply ligo/Makefile:18]: I don't think there is anything smarter, but in `Makefile` targets are supposed to be actual files produced by the corresponding commands https://www.gnu.org/software/make/manual/make.html#Rule-Introduction

> A target is usually the name of a file that is generated by a program; examples of targets are executable or object files. A target can also be the name of an action to carry out, such as ‘clean’ (see Phony Targets).

AFAIU there is no big advantage in non-phony targets, it just seems to be more conventional. Well, this convention is from the 20-th century and maybe nobody cares about it nowadays.

File: ligo/src/types.mligo
Comment (pasqu4le) [inline ligo/src/types.mligo:21]: It seems to be missing a `token_id` field, at least compared to its Lorentz counterpart.

File: ligo/src/types.mligo
Comment (pasqu4le) [inline ligo/src/types.mligo:29]: I see that there is no equivalent for the `Nonce` type, I suppose that's because it's less useful without the options to make `newtype`s in LIGO, but I'd argue even just a `nat` type synonym could be nice to have (as we do for the other ones).

File: ligo/src/types.mligo
Comment (pasqu4le) [inline ligo/src/types.mligo:77]: Even tho this is indeed part of the parameter, it's also part of FA2's interface for which there is a "group" above, do you think it would make sense to move this (and all the following types up to `fa2_parameter`) under the `FA2 types` banner?

File: ligo/src/types.mligo
Comment (pasqu4le) [inline ligo/src/types.mligo:126]:
```suggestion
type transfer_ownership_param = [@annot:newOwner] address
```
Again only to match the Lorentz counterpart, I think this should work.

File: ligo/src/types.mligo
Comment (pasqu4le) [inline ligo/src/types.mligo:128]:
```suggestion
type migrate_param = [@annot:newAddress] address
```

File: ligo/src/types.mligo
Comment (pasqu4le) [inline ligo/src/types.mligo:192]: I had written a couple of comments already but then realized this is a bit behind master and the PR needs a rebase.

Listing them here just in case they come useful, compared to the Lorentz version LIGO:
- lacks a `Drop_proposal` equivalent here in `parameter`
  (and as a consequence a corresponding entrypoint function)
- has `unit` as the argument of `Flush`, but it should actually be `nat option`

Maybe some of the comments above also are just about outdated stuff, sorry if that's the case.

**Review by pasqu4le — COMMENTED** [review]
One last comment about commit messages:

1. I'd remove the `of` from the first line `Start LIGO version of development`
2. `is far not trivial` should be corrected to `is far from trivial` or `is far from being trivial`
3. lacks a "problem/solution` description

File: ligo/Makefile
Comment (Martoon-00) [reply ligo/Makefile]: I updated the Makefile, now it does not rebuild the contract if it has already been built. Though the Makefile now looks a bit strange :roll_eyes: because unlike in stablecoin, I put the produced file in gitignored `out` folder, not directly at the `ligo` folder.

File: ligo/src/base_DAO.mligo
Comment (Martoon-00) [reply ligo/src/base_DAO.mligo:29]: This constructor will be removed soon or later because we will have an off-chain view instead, the constructor probably should not never in the LIGO version.

File: ligo/src/base_DAO.mligo
Comment (Martoon-00) [reply ligo/src/base_DAO.mligo:30]: Oh this is pretty weird, usually my editor handles trailing newlines. Gonna check what's going on.

File: ligo/src/permit.mligo
Comment (Martoon-00) [reply ligo/src/permit.mligo:1]: Yes, in Haskell this module contained only helpers, so there is nothing to put here yet.

File: ligo/src/types.mligo
Comment (Martoon-00) [reply ligo/src/types.mligo:6]: Since the only way for one module in LIGO to depend on another module is C-like inlining (literally inlining to form one big file), we can import the same module several times and have to deal with duplication, and the recommended way to do this is again C-like preprocessor magic. This is described e.g. on [this fancy page](https://tezosacademy.io/reason/chapter-preprocessor), and I think this applies to any syntax.

Usually, it is possible to inline the same module multiple times without problems, but in our case we have sum types in `types.mligo`, and so importing module twice ends with "conflicting constructor name".

File: ligo/src/types.mligo
Comment (Martoon-00) [reply ligo/src/types.mligo:21]: Could you point me to where do you see that? I can't find it :/

File: ligo/src/types.mligo
Comment (Martoon-00) [reply ligo/src/types.mligo:29]: Okay, added the necessary types to make Lorentz and LIGO versions match.

I only skip `DataToSign`, because IMO there is no sense in a non-polymorphic version of this type; I'll let the implementor of signing logic decide whether to use a record here or just pairs.

File: ligo/src/types.mligo
Comment (Martoon-00) [reply ligo/src/types.mligo:77]: Yes, this was a questionable point. Since this type is a dedicated parameter part, should I also move `fa2_parameter` above?

File: ligo/src/types.mligo
Comment (Martoon-00) [reply ligo/src/types.mligo:126]: Nope, it does not work :/

I suspect that LIGO cannot produce type annotations at all.

File: ligo/src/types.mligo
Comment (Martoon-00) [reply ligo/src/types.mligo:192]: Yes, these are just the very recent additions, should automatically be handled on rebase.

**Comment by Martoon-00** [comment]
One this that remains regarding all this is to decide what do we want to do with mirroring LIGO types?

At the moment I add the mirrored `Parameter`, `Storage` and `Config`.

The easiest option would be to use the new `ligoLayout` right in `Lorentz.Contract.BaseDAO.Types` and remove the mirrors. The shortcoming here is that such layout is suboptimal and operation size will increase - though I think that we don't care about operation size that much now when it is big anyway.

Another option is finishing the way I started, leave the mirrored types, and implement `IsoLigo` instances.

Probably this should be resolved already without me (unless we have a definite dicision today).

File: ligo/src/types.mligo
Comment (pasqu4le) [reply ligo/src/types.mligo:9]: Oh I see, so this is a "if it was imported already don't import it again" kinda thing. Got it

P.s. that's indeed a fancy page :sweat_smile:

File: ligo/src/types.mligo
Comment (pasqu4le) [reply ligo/src/types.mligo:24]: Oh you are right, there isn't one.

My bad, I got confused with `operator_param`,

File: ligo/src/types.mligo
Comment (pasqu4le) [reply ligo/src/types.mligo:85]: Not sure, I'm more for the "yes" only because being part of FA2 is more specific than being part of the parameter, but his is indeed subjective.

**Comment by pasqu4le** [comment]
> decide what do we want to do with mirroring LIGO types?

If I understand the problem I think we should consider "bending" the LIGO types to match the Lorentz ones (as they were before these changes) as much as possible.

I say because the `IsoLigo` solution seems excessively cumbersome and even tho we can now easily reproduce LIGO-compatible types in Lorentz we can also produce Lorentz-compatible types in LIGO (unless there's a problem I'm not seeing).

It would take longer (to modify the LIGO types) but this way we'd be able to avoid modifying the Lorentz one and we'd still have that reduced operation cost.

File: ligo/src/types.mligo
Comment (pasqu4le) [reply ligo/src/types.mligo:31]: I thought I had replied to this, but apparently not.

I was for moving that too anyway and it seems you have done so as well, so all good :slightly_smiling_face:

**Comment by pasqu4le** [comment]
This ended up requiring quite a bit more effort than expected.

I finally managed to make all the types between the LIGO version and the Haskell (for tests) version match after rebasing on the current `master`.

Note: I ended up keeping the `IsoLigo` approach, but using `ligoLayout` for the mirror types.

Everything should now be up-to-date and complete (except for the specification that I still have to write).
@rinn7e your review would be appreciated, I no longer expect to make changes here.

File: ligo/src/types.mligo
Comment (rinn7e) [inline ligo/src/types.mligo:60]:
```suggestion
type token_metadata_registry_param = address contract
```

**Review by rinn7e — APPROVED** [review]
Just a small typo. The rest looks good to me.

File: ligo/test/Ligo/BaseDAO/Types.hs
Comment (rinn7e) [inline ligo/test/Ligo/BaseDAO/Types.hs:237]: This is probably missed, the latest master now require `cProposalCheck` and `cRejectedProposalReturnValue                                                                             ` to have a storage type in the stack as well:   
https://github.com/tqtezos/baseDAO/blob/master/src/Lorentz/Contracts/BaseDAO/Types.hs#L89-L97

**Review by rinn7e — CHANGES_REQUESTED** [review]

File: ligo/test/Ligo/BaseDAO/Types.hs
Comment (rinn7e) [inline ligo/test/Ligo/BaseDAO/Types.hs:233]: `cDaoName` and `cDaoDescription` are only used in Lorentz autodoc feature. We probably do not need it here.

File: ligo/test/Ligo/BaseDAO/Types.hs
Comment (pasqu4le) [reply ligo/test/Ligo/BaseDAO/Types.hs]: :ok_hand: well spotted, added now

File: ligo/test/Ligo/BaseDAO/Types.hs
Comment (pasqu4le) [reply ligo/test/Ligo/BaseDAO/Types.hs]: Makes sense to me as well, removed now

File: ligo/src/types.mligo
Comment (rinn7e) [inline ligo/src/types.mligo:24]: I'm not exactly sure why. I was testing the fa2 ligo contract and it seems like it needs `layout:comb` here.

```
type operator =
  [@layout:comb]
  { owner : address
  ; operator : address
  }
```

So do we really need this or maybe I did something wrong.

Comment (pasqu4le) [reply ligo/src/types.mligo:25]: Yes that look correct. The test for the matching types passes in both cases because they are both `address`, but without `layout:comb` LIGO reorders them alphabetically, so they end up inverted (compared to how we expect them to be).
Fixed now

File: ligo/test/Ligo/BaseDAO/Types.hs
Comment (gromakovsky) [reply ligo/test/Ligo/BaseDAO/Types.hs]: Related: https://github.com/tqtezos/baseDAO/issues/72

File: .buildkite/pipeline.yml
Comment (pasqu4le) [reply .buildkite/pipeline.yml:67]: I haven't added CI test execution either because of this (they need the resulting contract), I also spent some time on it without luck and prefer to do so in a separate PR so that this one can be merged.

**Comment by pasqu4le** [comment]
Note: I simply added the specification in a separate file (`specification-ligo.md`) for now.

It should be referenced somewhere, but then again we also need to write a proper README for the LIGO version because there are too many differences to use the existing one, so I'll leave that to a future PR.

File: docs/specification-ligo.md
Comment (rinn7e) [inline docs/specification-ligo.md:38]:
```suggestion
In order to do so the contract has two type synonyms for `(string, bytes) map`
```

File: docs/specification-ligo.md
Comment (Martoon-00) [inline docs/specification-ligo.md:36]: Is it true for LIGO version?

File: docs/specification-ligo.md
Comment (rinn7e) [inline docs/specification-ligo.md:139]:
```suggestion
- Migration forms a two-step process:
  + After the migration starts, the contract `migration_status` state is set to `MigratingTo targetAddress` state.
  + After successful migration confirmation, `migration_status` is set to `MigratedTo newContractAddress` state. All the subsequent operations, except `transfer_contract_tokens`, will fail with the `MIGRATED` tag.
```

**Review by rinn7e — APPROVED** [review]
Just a few suggestions, the rest looks good from me.

File: docs/specification-ligo.md
Comment (pasqu4le) [reply docs/specification-ligo.md:36]: Fair point, changed it to:
```
BaseDAO is a concrete smart contract, but also a framework to implement various DAOs.
```

File: docs/specification-ligo.md
Comment (pasqu4le) [reply docs/specification-ligo.md]: Fixed

File: docs/specification-ligo.md
Comment (pasqu4le) [reply docs/specification-ligo.md]: Fixed here and in `specification.md` as well

**Pull request merged**
